### PR TITLE
feat(adapters): add Airtable adapter

### PR DESCRIPTION
**Pull request by kyen99**
## ☕️ Reasoning

This PR adds an adapter to use Airtable as a store for next-auth. Airtable is NOT a good long term solution for an authentication backend, but is useful for standing up a site with authentication FAST to test whether anyone wants to use [insert your probably terrible idea here].

## 🧢 Checklist

- [ X] Documentation
- [ X] Tests
- [ X] Ready to be merged

## 🎫 Affected issues

There are no existing tickets that mention Airtable as an adapter, but I intend to use this for many projects at my company for quickly testing ideas.

This doesn't fix any existing problem. (Unless someone secretly wanted an Airtable adapter.)

## 📌 Resources

- [Contributing guidelines](./CONTRIBUTING.md)
- [Code of conduct](./CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by kyen99** [comment]
Some additional notes I forgot to add to the PR...
- The tests require an airtable account and clone of this base to run: https://airtable.com/shr16Xd8glUk90c4P
- Airtable does not provide a way to programmatically create a database, so providing a public one with the correct schema that can be cloned seemed to be the best option (short of mocking the airtable sdk).

**Comment by kyen99** [comment]
If anyone comes across this PR while looking for an Airtable adapter, you can use this separate npm package until this is merged (or in case it is never merged).... https://www.npmjs.com/package/next-auth-adapter-airtable

File: packages/adapter-airtable/src/index.ts
Comment (balazsorban44) [inline packages/adapter-airtable/src/index.ts:4]:
```suggestion
export function AirtableAdapter(options: AirtableAdapterOptions): Adapter {
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thanks, do we really need so many files? :thinking: 

Code like this:

```ts
    async createUser(user: any) {
      return User.createUser(user)
    },

    async getUser(id) {
      return User.getUserById(id)
    },

    async getUserByEmail(email) {
      return User.getUserByEmail(email)
    },
```
looks like could be directly in the root `index` file instead. :+1:


So do we need some environment variables to be able to run the tests in the CI? is there a self-hosting/docker alternative? We cannot run tests on fork PRs otherwise. It might be acceptable, but I consider it a negative and would prefer not using env vars if possible.

We would need a documentation page under https://github.com/nextauthjs/next-auth/tree/main/docs/docs/adapters as well.

Make sure to mention this is not a good production solution, if relevant.

All in all, looks good, thank you!

**Comment by kyen99** [comment]
@balazsorban44 thank you for reviewing the PR!

I could move everything into a single file at the expense of a cognitive complexity of 30+ and about 200 lines of code (that's how I originally wrote it.) If this better matches the style of the existing codebase I will change to a single file.

There is no self-hosted airtable option. The only alternative is to mock the Airtable API. If that's preferable, I will do that.

I will start working on the documentation. Please let me know about the above two items.

Thank you!

**Comment by kyen99** [comment]
I combined this into fewer files as suggested by @balazsorban44. I believe everything has been addressed except for the tests requiring an airtable account and .env to store api key and base id.

Airtable doesn't provide any simulated environments for testing except a testDriver for their blocks SDK. I tried to use that testDriver, but it's not designed for testing the REST API. Mocking the library doesn't seem to make sense because of the way runBasicTests works (i.e. I would have to mock those calls as well).

I'll continue to think about other testing options. Maybe @ndom91 will have an idea?

File: packages/adapter-airtable/tests/index.test.ts
Comment (ThangHuuVu) [inline packages/adapter-airtable/tests/index.test.ts:9]: I think we'll need something like a docker setup to test this on CI 🤔

**Review by ThangHuuVu — COMMENTED** [review]
Thanks for the PR! 🙌 I've got some comments

File: packages/adapter-airtable/tests/index.test.ts
Comment (ndom91) [reply packages/adapter-airtable/tests/index.test.ts:9]: There's [NocoDB](https://github.com/nocodb/nocodb), a self-hostable open source Airtable alternative, but I don't think its API-compatible with airtable.

While your totally right here, I think as a closed source saas product it'll be really hard to pull off for Airtable. Happy to be proven wrong here though!

Comment (ndom91) [reply packages/adapter-airtable/tests/index.test.ts:9]: Another one I just came across while googling this, [Baserow](https://baserow.io/)

Comment (ThangHuuVu) [reply packages/adapter-airtable/tests/index.test.ts:9]: @ndom91 I agree, so what do you think, should we just ignore AirTable tests on CI then? 🤔

Comment (kyen99) [reply packages/adapter-airtable/tests/index.test.ts:9]: I reached out to the head of developer relations at Airtable to see if I can get some time from them to sort out how to handle testing. If they don't have a resolution that can be dockerized, I'll look into mocking the API calls. In the meantime, I'll resolve the items @ThangHuuVu pointed out above so this can be merged once testing is resolved.

Thanks everyone for supporting this!

File: packages/adapter-airtable/package.json
Comment (kyen99) [reply packages/adapter-airtable/package.json:29]: I removed this section.

File: packages/adapter-airtable/README.md
Comment (kyen99) [reply packages/adapter-airtable/README.md:8]: I removed the comment.

**Comment by ThangHuuVu** [comment]
hi @kyen99 sorry for the delay, do you think you can pick up the pieces and get this in soon? 🚢
About formatting, you can take a look at https://github.com/nextauthjs/next-auth/pull/6745 as a reference for:
- JSDoc for documentation (
- ESM output (check the `package.json` file)

Let me know if you have any questions!

**Pull request closed without merging**
